### PR TITLE
Initial port

### PR DESCRIPTION
**Pull request by gilescope**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        if name.ends_with("net") {
+            name.truncate(name.len() - 3);
+        }
+        // Capitalise
+        name.get_mut(0..1)
+            .expect("name should not be empty")
+            .make_ascii_uppercase();
+
+        let postfix = if is_reserved { "" } else { "Account" };
+        format!("{}{}", rust_valid_id(name), postfix)
```
Comment (bkchr) [inline src/lib.rs:132]: https://docs.rs/Inflector/0.11.4/inflector/cases/pascalcase/fn.to_pascal_case.html

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+                    "If no account specified, network should start `reserved` not:{}",
+                    name
+                );
+                name
+            },
+        );
+        let is_reserved = name.starts_with("reserved");
+        if name.ends_with("net") {
+            name.truncate(name.len() - 3);
+        }
```
Comment (bkchr) [inline src/lib.rs:125]: For what is this?

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+        let unreserved_networks: Vec<String> = self
+            .registry
+            .iter()
+            .filter_map(|r| r.network.clone())
+            .collect();
+        let unique_networks: HashSet<&String> = HashSet::from_iter(unreserved_networks.iter());
+        if unique_networks.len() != unreserved_networks.len() {
+            return Err("networks must be unique.".to_owned());
+        }
```
Comment (bkchr) [inline src/lib.rs:45]: This error should be more specific.

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+        .filter(|r| r.network.as_ref().filter(|r|r.starts_with("reserved")).is_some())
+        .map(|r| r.prefix)
+        .collect();
+
+    let number: Vec<_> = registry.registry.iter().map(|r| r.prefix).collect();
+    let count = registry.registry.len();
+    let name: Vec<_> = registry
+        .registry
+        .iter()
+        .map(|r| r.network.clone().unwrap_or_else(|| "Bare".to_string()))
```
Comment (bkchr) [inline src/lib.rs:182]: Bare?

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        .registry
+        .iter()
+        .filter(|r| r.network.as_ref().filter(|r|r.starts_with("reserved")).is_some())
+        .map(|r| format_ident!("{}", r.name()))
+        .collect();
+
+    let reserved_numbers: Vec<_> = registry
+        .registry
+        .iter()
+        .filter(|r| r.network.as_ref().filter(|r|r.starts_with("reserved")).is_some())
```
Comment (bkchr) [inline src/lib.rs:173]: Looks like you should introduce a function for `is_reserved`?

**Review by bkchr — COMMENTED** [review]
Please continue the rest in prs. Otherwise it is impossible to review this.

File: src/lib.rs (hunk 2 above)
Comment (gilescope) [reply src/lib.rs:125]: The json and the enum names don't quite match. If we're happy changing the enum names then I will happily ditch this.

File: src/lib.rs (hunk 4 above)
Comment (gilescope) [reply src/lib.rs:182]: There are three enum varients which don't have a network associated with them: BareSr25519, BareEd25519 and BareSecp256k1
Possibly there should be an is_bare() or has_network() function to identify this subset?

File: src/lib.rs (hunk 2 above)
Comment (bkchr) [reply src/lib.rs:125]: The point is that the enum is generated out of the json and not make the enum look the same as the current enum.

Comment (bkchr) [reply src/lib.rs:125]: So please remove it

File: src/lib.rs (hunk 4 above)
Comment (bkchr) [reply src/lib.rs:182]: Then we should set it to something in the json and make sure that this can not be null.

Comment (gilescope) [reply src/lib.rs:182]: fixed. Have set network to be BareEd25519 etc. for those ones.

**Pull request merged**